### PR TITLE
Add intersphinx mapping

### PR DESCRIPTION
**Pull request by danielhuppmann**
This PR adds the intersphinx mapping to pyam and restructures the docs (rst pages and code docstrings) to use intersphinx markup.

### CODE REVIEW

File: doc/source/getting_started.rst [hunk 1]
```diff
@@ -10,13 +12,13 @@ processing, which consists of renaming of model “native regions” and aggrega
 “common regions” used in a project.
 
 There are two main classes that the user interacts with when using the nomenclature
-package, **DataStructureDefinition** and **RegionProcessor** (a full list of all classes can be found in :ref:`api`). 
+package, :class:`DataStructureDefinition` and :class:`RegionProcessor`.
```
Comment (LauWien) [inline doc/source/getting_started.rst:15]: Ah that's nice that `DataStructureDefinition` is clickable.

**Review by phackstock — APPROVED** [review]
Very nice addition of referencing both within the package and to pyam as well.
Right now, the link for ":class:`RegionProcessor`" is not working because in the api docs I included it as "automodule:: nomenclature.processor.region" and not "autoclass:: RegionProcessor". Should we fix it here or should I address it in #76?

File: doc/source/getting_started.rst [hunk 2]
```diff
@@ -101,13 +103,13 @@ The following outlines how to use the nomenclature package:
...
-  containing yaml files. See :ref:`dir-structure` for details. 
+* :class:`DataStructureDefinition` and :class:`RegionProcessor` are initialized from
+  directories containing yaml files. See :ref:`dir-structure` for details.
 
-* The processor argument of *process* is optional and may only to be used if there are  
-  model mappings. See :ref:`toplevel-functions` for details.
+* The processor argument of func:`process` is optional and may only to be used if there
+  are model mappings. See :ref:`toplevel-functions` for details.
 
 * If not all dimensions of the **DataStructureDefinition** should be validated, a
```
Comment (LauWien) [inline doc/source/getting_started.rst:114]:
```suggestion
* If not all dimensions of the :class:`DataStructureDefinition` should be validated, a
```
In the "Notes" of "Directory structure for definitions and mappings" the classes DataStructureDefinition and RegionProcessor could also be referred as :class:`DataStructureDefinition` and :class:`RegionProcessor` as implemented here if wanted.

File: nomenclature/core.py [hunk 3]
```diff
@@ -28,20 +28,20 @@ def process(
...
-    df : pyam.IamDataFrame
-        Input data to be validated and aggregated.
-    dsd : DataStructureDefinition
+    df : :class:`pyam.IamDataFrame`
+        Scenario data to be validated and aggregated.
+    dsd : :class:`DataStructureDefinition`
         Codelists that are used for validation.
     dimensions : list, optional
         Dimensions to be used in the validation, defaults to all dimensions defined in
         *dsd*
```
Comment (LauWien) [inline nomenclature/core.py:37]:
```suggestion
        `dsd`
```

File: nomenclature/core.py [hunk 4]
```diff
@@ -28,20 +28,20 @@ def process(
...
-    processor : RegionProcessor, optional
-        Region processor that will perform region renaming and aggregation if provided
+    processor : :class:`RegionProcessor`, optional
+        Region processor to perform region renaming and aggregation (if given)
```
Comment (LauWien) [inline nomenclature/core.py:39]: "(if given)" possible redundant since `processor` is "optional"

**Review by LauWien — APPROVED** [review]
Looks great :) 
Just a few suggestions inline.

**Comment by danielhuppmann** [comment]
Response to @phackstock:

> Right now, the link for ":class:RegionProcessor" is not working because in the api docs I included it as "automodule:: nomenclature.processor.region" and not "autoclass:: RegionProcessor". Should we fix it here or should I address it in #76?

I think a priority is to make intersphinx operate smoothly also for other users pointing to our docs, so we should make sure that :class:`nomenclature.RegionProcessor` just works. So we should restructure the API documentation (ideally as a separate later PR - it's not a big problem if the link is not created right now).

**Comment by phackstock** [comment]
@danielhuppmann ok, sounds good. Then good to merge from my side. I'll rebase #76 once we've merged.

File: nomenclature/core.py (hunk 4 above)
Comment (danielhuppmann) [reply nomenclature/core.py:39]: Yes, redundant, but on purpose, because users may overlook the "optional" tag

**Pull request merged**